node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java

```diff
@@ -93,31 +94,78 @@ void updateProvisioningNodes(NodeList nodes, Mutex lock) {
         });
     }
 
-    void convergeToCapacity(NodeList nodes) {
-        Collection<Node> removableHosts = getRemovableHosts(nodes);
-        List<NodeResources> preProvisionCapacity = preprovisionCapacityFlag.value().stream()
-                .flatMap(cap -> {
-                    NodeResources resources = new NodeResources(cap.getVcpu(), cap.getMemoryGb(), cap.getDiskGb(), 1);
-                    return IntStream.range(0, cap.getCount()).mapToObj(i -> resources);
-                })
-                .sorted(NodeResourceComparator.memoryDiskCpuOrder().reversed())
-                .collect(Collectors.toList());
+    /** Converge zone to wanted capacity */
+    private void convergeToCapacity(NodeList nodes) {
+        List<Node> removableHosts = removableHostsOf(nodes);
+        List<Node> excessHosts;
+        if (nodeRepository().zone().getCloud().dynamicProvisioning()) {
+            excessHosts = preprovisionCapacity(removableHosts);
+        } else {
+            excessHosts = provisionExactCapacity(removableHosts);
+        }
+        excessHosts.forEach(host -> {
+            try {
+                hostProvisioner.deprovision(host);
+                nodeRepository().removeRecursively(host, true);
+            } catch (RuntimeException e) {
+                log.log(Level.WARNING, "Failed to deprovision " + host.hostname() + ", will retry in " + interval(), e);
+            }
+        });
+    }
 
+    /** Provision the exact capacity declared for this zone, if any.
+     *
+     * @return Excess hosts that can be deprovisioned.
+     */
+    private List<Node> provisionExactCapacity(List<Node> removableHosts) {
+        List<NodeResources> capacity = capacityFrom(exactCapacityFlag);
+        if (capacity.isEmpty()) return List.of();
+
+        List<Node> excessHosts = new ArrayList<>(removableHosts);
+        List<NodeResources> unsatisfiedCapacity = new ArrayList<>(capacity);
+        for (var resources : capacity) {
+            for (var host : removableHosts) {
```

Comment:
Sort by cost, otherwise you risk counting a large host as satisfying a small HostCapacity, but then you wont find match for a larget HostCapacity request